core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+class SnapshotOperations implements Serializable {
+
+  private List<Snapshot> snapshots;
+  private Supplier<List<Snapshot>> snapshotSupplier;
+  private Map<Long, Snapshot> snapshotsById;
+  private Map<String, SnapshotRef> refs;
+
+  SnapshotOperations() {
+    snapshots = Lists.newArrayList();
+    snapshotsById = Maps.newHashMap();
+    refs = Maps.newHashMap();
+  }
+
+  SnapshotOperations(List<Snapshot> snapshots, Map<String, SnapshotRef> refs) {
+    this(snapshots, null, refs);
+  }
+
+  SnapshotOperations(SnapshotOperations snapshotOperations) {
+    this(
+        snapshotOperations.snapshots, snapshotOperations.snapshotSupplier, snapshotOperations.refs);
+  }
+
+  SnapshotOperations(
+      List<Snapshot> snapshots,
+      Supplier<List<Snapshot>> snapshotSupplier,
+      Map<String, SnapshotRef> refs) {
+    this.snapshots = Lists.newArrayList(snapshots);
+    this.snapshotSupplier = snapshotSupplier;
+    this.refs = Maps.newHashMap(refs);
+
+    indexSnapshots();
+  }
+
+  List<Snapshot> snapshots() {
+    ensureLoaded();
+
+    return Collections.unmodifiableList(snapshots);
+  }
+
+  Snapshot snapshot(long id) {
+    if (!snapshotsById.containsKey(id)) {
```

Comment:
I'm not sure this is correct. `ensureLoaded()` doesn't populate anything to `snapshotsById`